src/tools.js

```diff
@@ -0,0 +1,323 @@
+const Apify = require('apify');
+const _ = require('underscore');
+const utils = require('apify-shared/utilities');
+const { DEFAULT_VIEWPORT, BASE_URL_LABEL, OUTPUT_COLORS, STATUS_CODES, URL_PREFIX_REGEX } = require("./consts");
+
+const { utils: { log } } = Apify;
+
+const setDefaultViewport = (_pageId, launchContext) => {
+    launchContext.launchOptions.defaultViewport = DEFAULT_VIEWPORT;
+};
+
+/**
+ * This function normalizes the URL and removes the #fragment.
+ * @param {string} url 
+ * @returns {string} normalized url
+ */
+const normalizeUrl = (url) => {
+    const nurl = utils.normalizeUrl(url);
+    if (nurl) return nurl;
+
+    const index = url.indexOf('#');
+    if (index > 0) return url.substring(0, index);
+
+    return url;
```

Comment:
Not my function (just moved from `main.js`) but I guess it doesn't want to throw away the original url even when it can't be normalized by `utils.normalizeUrl` function. I found snippet of `utils.normalizeUrl` in which `null` is returned as a normalized url:

```javascript
if (typeof url !== 'string' || !url.length) {
    return null;
}

const urlObj = parseUrl(url.trim());
if (!urlObj.protocol || !urlObj.host) {
    return null;
}
```

I'm not sure if both `protocol` and `host` should be required and whether the original url should be marked bad in case the're missing.